### PR TITLE
add BFloat16 operators on CPU: nextafter, atan2

### PR DESCRIPTION
**Pull request by jiayisunx**
Added BFloat16 support for atan2, and optimized BFloat16 support for nextafter on CPU, and collected the benchmark data of the two OPs(atan2, nextafter) for BFloat16 and Float32 data type by using the operator_benchmark tool of PyTorch on the platform of Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz

Number of cores: 1 core, 28 cores(1 socket)
[atan2_benchmark.txt](https://github.com/pytorch/pytorch/files/7028984/atan2_benchmark.txt)
[nextafter_benchmark.txt](https://github.com/pytorch/pytorch/files/7028985/nextafter_benchmark.txt)

### CODE REVIEW

**Review by kshitij12345 — CHANGES_REQUESTED** [review]
Thanks @jiayisunx!

Please update the test for `nextafter` bfloat16 (to check the output on tensor with bigger size) so that the vectorized path is also verified.

For `atan2` adding ref to its opinfo entry would allow us to verify the correctness of the implementation
https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/torch/testing/_internal/common_methods_invocations.py#L5723-L5724

Sample of adding `ref`:
https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/torch/testing/_internal/common_methods_invocations.py#L5455-L5457

Relevant test after adding that field would be
https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/test/test_ops.py#L194-L199

**Comment by jiayisunx** [comment]
> Thanks @jiayisunx!
> 
> Please update the test for `nextafter` bfloat16 (to check the output on tensor with bigger size) so that the vectorized path is also verified.
> 
> For `atan2` adding ref to its opinfo entry would allow us to verify the correctness of the implementation
> https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/torch/testing/_internal/common_methods_invocations.py#L5723-L5724
> 
> Sample of adding `ref`:
> https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/torch/testing/_internal/common_methods_invocations.py#L5455-L5457
> 
> Relevant test after adding that field would be
> https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/test/test_ops.py#L194-L199

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -3947,7 +3947,10 @@ def sample_inputs_nextafter(op_info, device, dtype, requires_grad, **kwargs):
     cases = (
         ((S, S), (S, S), False),
         ((S, S), (S,), False),
-        ((S, ), (S, S), True)
+        ((S, ), (S, S), True),
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:3950]: You should undo the changes here.

Sorry I forgot to mention, you will have to update the following test (since NumPy doesn't have nextafter which supports bfloat16, we had to hand-roll a test with golden values)

https://github.com/pytorch/pytorch/blob/913c1f83f49f9e1e2a494186cc0069d780cee852/test/test_binary_ufuncs.py#L1802-L1804

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h [hunk 2]
```diff
@@ -431,6 +431,15 @@ template <> class Vectorized<BFloat16> {
     auto o2 = _mm256_xor_ps(mask, hi);
     return cvtfp32_bf16(o1, o2);
   }
+  Vectorized<BFloat16> nextafter(const Vectorized<BFloat16> &b) const {
```
Comment (kshitij12345) [inline aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:434]: Btw, I am curious to see if it works.

(I had tried casting to float, doing `nextafter` and casting back to bfloat16 while adding non-vectorized support). 
That did not work as the value of `nextafter` in float is much smaller than the precision of `bfloat16` and ends up being discarded when converted back to `bfloat16`.
Or
Maybe I did something incorrectly

**Comment by kshitij12345** [comment]
Actually staring at
https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/test/test_ops.py#L194-L199

I can see that we don't do reference testing for `bfloat16` (probably since numpy doesn't support it). I think we should add a test in `test_torch.py` for `atan2`'s bfloat16 support.

Apologies for the confusion @jiayisunx.

@mruberry what do you think? 
Also, should we look at updating reference test to test for `bfloat16` in a seperate PR?

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (jiayisunx) [reply torch/testing/_internal/common_methods_invocations.py]: Done, thanks.

File: aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h (hunk 2 above)
Comment (jiayisunx) [reply aten/src/ATen/cpu/vec/vec256/vec256_bfloat16.h:434]: There was no difference in the precision of bfloat16 between non-vectorized supporting and vectorized supporting.

**Comment by jiayisunx** [comment]
> Actually staring at
> https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/test/test_ops.py#L194-L199
> 
> I can see that we don't do reference testing for `bfloat16` (probably since numpy doesn't support it). I think we should add a test in `test_torch.py` for `atan2`'s bfloat16 support.
> 
> Apologies for the confusion @jiayisunx.
> 
> @mruberry what do you think?
> Also, should we look at updating reference test to test for `bfloat16` in a seperate PR?

**Comment by mruberry** [comment]
> Actually staring at
> https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/test/test_ops.py#L194-L199
> 
> I can see that we don't do reference testing for `bfloat16` (probably since numpy doesn't support it). I think we should add a test in `test_torch.py` for `atan2`'s bfloat16 support.
> 
> Apologies for the confusion @jiayisunx.
> 
> @mruberry what do you think?
> Also, should we look at updating reference test to test for `bfloat16` in a seperate PR?

Updating reference testing to include bfloat16 sounds fine as long as we don't hit numerical stability issues when doing so. That test should be very reliable and not require special handling since it's run on every opinfo.

But yes, both nextafter and atan2 should have existing tests comparing them with their NumPy counterparts outside of opinfo testing, and they should be extended with bfloat16 checking OR a new bfloat16 test using golden values added

**Comment by jiayisunx** [comment]
> > Actually staring at
> > https://github.com/pytorch/pytorch/blob/0a66d5b3253fd2d2304f3897526db3c8fb139376/test/test_ops.py#L194-L199
> > 
> > I can see that we don't do reference testing for `bfloat16` (probably since numpy doesn't support it). I think we should add a test in `test_torch.py` for `atan2`'s bfloat16 support.
> > Apologies for the confusion @jiayisunx.
> > @mruberry what do you think?
> > Also, should we look at updating reference test to test for `bfloat16` in a seperate PR?
> 
> Updating reference testing to include bfloat16 sounds fine as long as we don't hit numerical stability issues when doing so. That test should be very reliable and not require special handling since it's run on every opinfo.
> 
> But yes, both nextafter and atan2 should have existing tests comparing them with their NumPy counterparts outside of opinfo testing, and they should be extended with bfloat16 checking OR a new bfloat16 test using golden values added

Have added bfloat16 checking, thanks.

File: test/test_binary_ufuncs.py [hunk 3]
```diff
@@ -1842,6 +1842,12 @@ def test_nextafter_bfloat16(self, device, dtype):
             actual = torch.nextafter(from_t, to_t).item()
             self.assertEqual(actual, expected, atol=0, rtol=0)
 
+        a = torch.randn(100, device=device, dtype=dtype)
+        b = torch.randn(100, device=device, dtype=dtype)
+        actual = torch.nextafter(a, b)
+        expected = np.nextafter(a.cpu().float().numpy(), b.cpu().float().numpy())
+        self.assertEqual(actual.float(), expected, atol=0.01, rtol=0.01)
```
Comment (kshitij12345) [inline test/test_binary_ufuncs.py:1849]: This looks incorrect, (it passes because we do the same operations for numpy such that they both agree)

```python
>>> import torch
>>> import numpy as np
>>> z = torch.zeros(())
>>> z = torch.zeros((), dtype=torch.bfloat16)
>>> o = torch.ones((), dtype=torch.bfloat16)
>>> torch.nextafter(z, o)
tensor(9.1835e-41, dtype=torch.bfloat16)
>>> torch.nextafter(z.float(), o.float())
tensor(1.4013e-45)
>>> np.nextafter(z.float().numpy(), o.float().numpy())
1e-45
```

I think the correct way will be to extend the above golden value test to also check for larger input as here.
```python
for from_v, to_v, expected in cases: 
      # scalar path
     from_t = torch.tensor([from_v], device=device, dtype=dtype) 
     to_t = torch.tensor([to_v], device=device, dtype=dtype) 
     actual = torch.nextafter(from_t, to_t).item()
     self.assertEqual(actual, expected, atol=0, rtol=0) 
    
     # vector path
     from_t = torch.tensor([from_v] * 100, device=device, dtype=dtype) 
     to_t = torch.tensor([to_v] * 100, device=device, dtype=dtype) 
     actual_t = torch.nextafter(from_t, to_t).item()
     expected_t = torch.tensor([expected] * 100, device=device, dtype=dtype) 
     self.assertEqual(actual_t, expected_t, atol=0, rtol=0) 
    

```
